File: Wordle_Java/src/PlayerInfo.java
/*
 * Wordle program
 * CS 203 Project 3
 * Benjamin Hall
 * 2022.03.06
 */

import java.io.*;
import java.util.*;
import java.util.function.*;
import java.util.stream.*;

/**
 * A class to contain information about a player.
 * 
 * @author Benjamin Hall
 */
public class PlayerInfo {
    private final String _username;
    private final HashSet<String> _wordsPlayed;
    private final int[] _numGuesses;
    private int _maxWinStreak;
    private int _curWinStreak;

    /**
     * Creates an object for a new player.
     * 
     * @param username
     *        The username of the player
     */
    public PlayerInfo(String username) {
        this(username, new HashSet<String>(), new int[6], 0, 0);
    }

    /**
     * Creates an object containing information about an existing player.
     * 
     * @param username
     *        The username of the player
     * @param wordsPlayed
     *        A set of words the player has already played
     * @param numGuesses
     *        An array containing the number of times the
     *        player has guessed a word in (i + 1) attempts,
     *        where i is the index of the array from [0, 6)
     * @param maxWinStreak
     *        The maximum winning streak
     * @param curWinStreak
     *        The current winning streak
     */
    public PlayerInfo(
        String username,
        HashSet<String> wordsPlayed,
        int[] numGuesses,
        int maxWinStreak,
        int curWinStreak
    ) {
        _username = username;
        _wordsPlayed = wordsPlayed;
        _numGuesses = numGuesses;
        _maxWinStreak = maxWinStreak;
        _curWinStreak = curWinStreak;
    }

    /**
     * Returns the username of this player.
     */
    public String getUsername() {
        return _username;
    }

    /**
     * Gets a random, unplayed word for a new Wordle game.
     * 
     * @param dictionary
     *        The dictionary from which to find a word
     * @return
     *        A random word the player has not yet played
     */
    public String getRandomWord(Set<String> dictionary) {
        int unplayedWordsCount = dictionary.size() - _wordsPlayed.size();
        int randomWordIdx = (int)(Math.random() * unplayedWordsCount);
        return dictionary.stream()
            .filter(Predicate.not(_wordsPlayed::contains))
            .skip(randomWordIdx)
            .findFirst().get()
            .toUpperCase();
    }

    /**
     * Adds a word that the player successfully guessed
     * to the database.
     * 
     * @param word
     *        The word the player successfully guessed
     * @param numGuesses
     *        The number of guesses the player made
     */
    public void addWonWord(String word, int numGuesses) {
        _wordsPlayed.add(word);
        ++_numGuesses[numGuesses - 1];
        ++_curWinStreak;
        _maxWinStreak = Math.max(_maxWinStreak, _curWinStreak);
    }

    /**
     * Adds a word that the player failed to guess
     * to the database.
     * 
     * @param word
     *        The word the player failed to guess
     */
    public void addLostWord(String word) {
        _wordsPlayed.add(word);
        _curWinStreak = 0;
    }

    /**
     * Returns a formatted string with the player's
     * current statistics.
     */
    public String getStats() {
        StringBuilder builder = new StringBuilder();
        builder.append("Number of Words Played: ")
            .append(_wordsPlayed.size())
            .append(System.lineSeparator());

        int winRate;
        if (_wordsPlayed.size() == 0) {
            winRate = 0;
        } else {
            int totalNumGuesses = Arrays.stream(_numGuesses).sum();
            winRate = (int)Math.round(100.0 * totalNumGuesses / _wordsPlayed.size());
        }
        builder.append("Win Rate: ")
            .append(winRate).append("%")
            .append(System.lineSeparator());

        builder.append("Current Win Streak: ")
            .append(_curWinStreak)
            .append(System.lineSeparator());

        builder.append("Maximum Win Streak: ")
            .append(_maxWinStreak)
            .append(System.lineSeparator());

        builder.append("Guess Distribution:");
        OptionalInt maxNumGuesses = Arrays.stream(_numGuesses).max();

        /* print the guess distribution with bars, max 12 */
        double barFactor;
        if (maxNumGuesses.isPresent()) {
            barFactor = 12.0 / maxNumGuesses.getAsInt();
        } else {
            barFactor = 0;
        }

        for (int i = 0; i < _numGuesses.length; ++i) {
            int numBars = (int)Math.round(barFactor * _numGuesses[i]);
            String bars = "=".repeat(numBars);

            builder.append(System.lineSeparator())
                .append(i + 1).append(": ")
                .append(bars).append(" ")
                .append(_numGuesses[i]);
        }

        return builder.toString();
    }

    /**
     * Returns a String representation of this player's
     * information for saving to a database.
     */
    @Override
    public String toString() {
        StringBuilder builder = new StringBuilder();
        builder.append("Username: ")
            .append(_username)
            .append(System.lineSeparator());

        builder.append("Words Played: ")
            .append(String.join(",", _wordsPlayed))
            .append(System.lineSeparator());

        String numGuessesStr = Arrays.stream(_numGuesses)
            .mapToObj(Integer::toString)
            .collect(Collectors.joining(","));
        builder.append("Number of Guesses: ")
            .append(numGuessesStr)
            .append(System.lineSeparator());

        builder.append("Maximum Win Streak: ")
            .append(_maxWinStreak)
            .append(System.lineSeparator());

        builder.append("Current Win Streak: ")
            .append(_curWinStreak);

        return builder.toString();
    }

    /**
     * Writes this player's information to a file.
     * 
     * @param fileName
     *        The name of the file
     * 
     * @throws IOException
     *        if the data could not be written to the file
     */
    public void writeToFile(String fileName) throws IOException {
        File file = new File(fileName);
        /* make sure file exists */
        file.createNewFile();

        BufferedWriter writer = new BufferedWriter(new FileWriter(file));
        writer.append(toString());
        writer.close();
    }

    /**
     * Parses a player's information from a file and creates a new
     * object containing the parsed information.
     * 
     * @param fileName
     *        The name of the player file
     * @return
     *        An object containing information about the player
     * 
     * @throws IOException
     *        if the data could not be read from the file
     */
    public static PlayerInfo createFromFile(String fileName) throws IOException, NumberFormatException {
        File file = new File(fileName);
        if (!file.exists()) {
            /* the database does not yet exist */
            return null;
        }

        /* read all the lines in the file */
        ArrayList<String> linesInFile = new ArrayList<String>();
        {
            Scanner reader = new Scanner(file);
            while (reader.hasNextLine()) {
                linesInFile.add(reader.nextLine());
            }
            reader.close();
        }

        if (linesInFile.size() != 5) {
            /* corrupt database file */
            throw new IOException("Error: corrupt player database file: " + fileName);
        }

        /* parse the lines in the file */
        DatabaseEntry<String> username = PlayerDatabaseParser.parseLine(linesInFile.get(0), s -> s);
        DatabaseEntry<HashSet<String>> wordsPlayed = PlayerDatabaseParser.parseSet(linesInFile.get(1), s -> s, HashSet::new);
        DatabaseEntry<ArrayList<Integer>> numGuessesList = PlayerDatabaseParser.parseList(linesInFile.get(2), Integer::parseInt, ArrayList::new);
        DatabaseEntry<Integer> maxWinStreak = PlayerDatabaseParser.parseLine(linesInFile.get(3), Integer::parseInt);
        DatabaseEntry<Integer> curWinStreak = PlayerDatabaseParser.parseLine(linesInFile.get(4), Integer::parseInt);

        /* parse the number of guesses into an array */
        int[] numGuesses = new int[6];
        for (int i = 0; i < numGuesses.length && i < numGuessesList.value.size(); ++i) {
            numGuesses[i] = numGuessesList.value.get(i);
        }

        return new PlayerInfo(username.value, wordsPlayed.value, numGuesses, maxWinStreak.value, curWinStreak.value);
    }
}

File: Wordle_CS/Players/PlayerInfo.cs
using System.Text;

namespace Wordle_CS.Players;

/// <summary>
/// Contains information about a Wordle player.
/// <para/>
/// A player has a:
/// <list type="bullet">
///     <item>username</item>
///     <item>list of words played</item>
///     <item>guess distribution</item>
///     <item>maximum win streak</item>
///     <item>current win streak</item>
/// </list>
/// </summary>
public class PlayerInfo
{
    /// <summary>
    /// This player's username.
    /// </summary>
    public string Username { get; init; }
    private readonly HashSet<string> wordsPlayed;
    private readonly ulong[] numGuesses;
    private ulong maxWinStreak;
    private ulong curWinStreak;

    private readonly Random rnd = new();

    /// <summary>
    /// Initializes data for a new player.
    /// </summary>
    /// <param name="username">The username of the player</param>
    public PlayerInfo(string username) : this(username, new(), new ulong[6], 0, 0)
    { }

    /// <summary>
    /// Loads data for an existing player.
    /// </summary>
    /// <param name="username">The username of the player</param>
    /// <param name="wordsPlayed">A set of words the player has already played</param>
    /// <param name="numGuesses">
    ///     An array containing the number of times the
    ///     player has guessed a word in (i + 1) attempts,
    ///     where i is the index of the array from [0, 6)
    /// </param>
    /// <param name="maxWinStreak">The maximum winning streak</param>
    /// <param name="curWinStreak">The current winning streak</param>
    public PlayerInfo(string username, HashSet<string> wordsPlayed, ulong[] numGuesses, ulong maxWinStreak, ulong curWinStreak)
    {
        Username = username;
        this.wordsPlayed = wordsPlayed;
        this.numGuesses = numGuesses;
        this.maxWinStreak = maxWinStreak;
        this.curWinStreak = curWinStreak;
    }

    /// <summary>
    /// Gets a random word the player has not yet played.
    /// </summary>
    /// <param name="dictionary">The dictionary from which to find a word</param>
    public string GetRandomWord(HashSet<string> dictionary)
    {
        var unplayedWordsCnt = dictionary.Count - wordsPlayed.Count;
        var randomWordIdx = rnd.Next(0, unplayedWordsCnt);
        return dictionary.Where(w => !wordsPlayed.Contains(w)).Skip(randomWordIdx).First();
    }

    /// <summary>
    /// Adds a word the player has successfully guessed to their database.
    /// <para/>
    /// This function increments the current win streak, adds
    /// the guess to the list of words played and to the guess
    /// distribution, and updates the max win streak if appropriate.
    /// </summary>
    /// <param name="word">The word the player successfully guessed</param>
    /// <param name="numGuesses">The number of guesses the player made</param>
    public void AddWonWord(string word, ulong numGuesses)
    {
        wordsPlayed.Add(word);
        ++this.numGuesses[numGuesses - 1];
        ++curWinStreak;
        maxWinStreak = Math.Max(maxWinStreak, curWinStreak);
    }

    /// <summary>
    /// Adds a word the player has failed to guess to their database.
    /// <para/>
    /// This function resets the current win streak to 0 and adds
    /// the guess to the list of words played. The number of guesses
    /// is not added to the player's guess distribution.
    /// </summary>
    /// <param name="word">The word the player failed to guess</param>
    public void AddLostWord(string word)
    {
        wordsPlayed.Add(word);
        curWinStreak = 0;
    }

    /// <summary>
    /// Returns a string representation of this player's information.
    /// </summary>
    /// <returns></returns>
    public override string ToString()
    {
        StringBuilder sb = new();
        sb.AppendLine($"Username: {Username}");
        sb.AppendLine($"Words Played: {string.Join(",", wordsPlayed)}");
        sb.AppendLine($"Number of Guesses: {string.Join(",", numGuesses)}");
        sb.AppendLine($"Maximum Win Streak: {maxWinStreak}");
        sb.AppendLine($"Current Win Streak: {curWinStreak}");
        return sb.ToString();
    }

    /// <summary>
    /// Returns a string with formatted player statistics.
    /// <para/>
    /// Player statistics consist of:
    /// <list type="bullet">
    ///     <item>Number of words played</item>
    ///     <item>Win rate</item>
    ///     <item>Current and max win streak</item>
    ///     <item>Guess distribution</item>
    /// </list>
    /// </summary>
    public string GetStats()
    {
        StringBuilder sb = new();
        sb.AppendLine($"Number of Words Played: {wordsPlayed.Count}");

        var winRate = wordsPlayed.Count == 0 ? 0 : (uint)Math.Round(100.0 * numGuesses.Aggregate((sum, x) => sum + x) / wordsPlayed.Count);
        sb.AppendLine($"Win Rate: {winRate}%");

        sb.AppendLine($"Current Win Streak: {curWinStreak}");
        sb.AppendLine($"Maximum Win Streak: {maxWinStreak}");

        sb.AppendLine("Guess Distribution:");
        var maxNumGuesses = numGuesses.Max();
        var barFactor = maxNumGuesses == 0 ? 0 : 12.0 / maxNumGuesses;

        foreach (var (numGuess, i) in numGuesses.Select((numGuess, i) => (numGuess, i)))
        {
            var numBars = (int)Math.Round(barFactor * numGuess);
            var bars = string.Concat(Enumerable.Repeat("=", numBars));
            sb.AppendLine($"{i + 1}: {bars} {numGuess}");
        }

        return sb.ToString().Trim();
    }

    /// <summary>
    /// Writes a player's data to a file.
    /// </summary>
    /// <param name="filename">The file to which to write</param>
    public void WriteToFile(string filename)
    {
        using var file = File.Create(filename);
        using BufferedStream buffer = new(file);
        using StreamWriter writer = new(buffer);
        writer.Write(ToString());
    }

    /// <summary>
    /// Reads a player's information from a file.
    /// <para/>
    /// The function throws an exception on parse errors.
    /// </summary>
    /// <param name="filename">The file from which to read</param>
    /// <returns>
    ///     A player's information if the operation
    ///     succeeded, or a nullopt
    /// </returns>
    public static PlayerInfo? CreateFromFile(string filename)
    {
        string fileContents;
        try
        {
            using var file = File.OpenRead(filename);
            using BufferedStream buffer = new(file);
            using StreamReader reader = new(buffer);
            fileContents = reader.ReadToEnd();
        }
        catch (Exception)
        {
            return null;
        }

        return FromStr(fileContents, new IOException($"Error: corrupt player database file: {filename}"));
    }

    /// <summary>
    /// Reads a player's info from a string.
    /// <para/>
    /// The function throws an exception on parse errors.
    /// </summary>
    /// <param name="playerData">The string to parse</param>
    /// <param name="badDataErr">The exception to throw on parse errors</param>
    /// <returns>
    ///     A player's information if the operation
    ///     succeeded, or a nullopt
    /// </returns>
    private static PlayerInfo FromStr(string playerData, IOException badDataErr)
    {
        var linesInFile = playerData.Lines().ToList();
        if (linesInFile.Count != 5)
        {
            throw badDataErr;
        }

        var username = DatabaseEntry<string>.FromLine(linesInFile[0], s => s) ?? throw badDataErr;
        var wordsPlayed = DatabaseEntry<string>.FromCollection<HashSet<string>>(linesInFile[1], s => s) ?? throw badDataErr;
        var numGuessesList = DatabaseEntry<ulong>.FromCollection<List<ulong>>(linesInFile[2], ulong.Parse) ?? throw badDataErr;
        var maxWinStreak = DatabaseEntry<ulong>.FromLine(linesInFile[3], ulong.Parse) ?? throw badDataErr;
        var curWinStreak = DatabaseEntry<ulong>.FromLine(linesInFile[4], ulong.Parse) ?? throw badDataErr;

        var numGuesses = new ulong[6];
        foreach (var (numGuess, i) in numGuessesList.Value.Take(numGuesses.Length).Select((n, i) => (n, i)))
        {
            numGuesses[i] = numGuess;
        }

        return new PlayerInfo(username.Value, wordsPlayed.Value, numGuesses, maxWinStreak.Value, curWinStreak.Value);
    }
}
